cpp/ql/src/Likely Bugs/Underspecified Functions/ImplicitFunctionDeclaration.ql

```diff
@@ -0,0 +1,41 @@
+/**
+ * @name Implicit function declaration
+ * @description An implicitly declared function is assumed to take no
+ * arguments and return an integer. If this assumption does not hold, it
+ * may lead to unpredictable behavior.
+ * @kind problem
+ * @problem.severity warning
+ * @precision medium
+ * @id cpp/implicit-function-declaration
+ * @tags correctness
+ *       maintainability
+ */
+
+import cpp
+import MistypedFunctionArguments
+import TooFewArguments
+import TooManyArguments
+import semmle.code.cpp.commons.Exclusions
+
+predicate sameLocation(Location loc1, Location loc2) {
+  loc1.getFile() = loc2.getFile() and
+  loc1.getStartLine() = loc2.getStartLine() and
+  loc1.getStartColumn() = loc2.getStartColumn()
+}
```

Comment:
This kind of predicate can be very expensive to compute, as it's naturally a cartesian product (i.e. an N x N table of data with no way to directly compute `loc1` or `loc2` from the other).  If you haven't noticed a problem on some of our bigger and more problematic projects (like wireshark) it's (reasonably) likely that the optimizer is doing a good job of dealing with this and we can leave it as it is.

Reply:
You could get an explosion here if many locations have the same starting point but different end points. If two locations are identical on all five columns, the extractor will arrange that they are the same `Location` object, except if they are locations for different "kinds" of things: expressions, statements, and "default", I think. So if you join on end line+column too, you can get at most 3 x 3 extra rows for such a location. You have to check that the optimiser joins on all five columns in one step, or else it's going to explode on some snapshot.

To ensure good performance, use [folding](https://help.semmle.com/QL/learn-ql/writing-queries/debugging-queries.html#fold-predicates). If that's not good enough, you can restrict one side of the five-column join to be calls only and the other side to be implicit declaration entries only.

Reply:
The declaration generated seems to have an unreliable 'end' position, unfortunately - sometimes the declaration is 1 character long (i.e., the 'end' column is one more than the 'start' column), and other times its *almost* the same end position as the function call. I initially joined on both 'start' and 'end', but that meant that the query couldn't always find the declaration generated by the call.

I did notice a couple of cartesian products in the RA, which I couldn't quite make sense of. I guess it's due to this predicate - I will see if I can fold my way out of them!

Reply:
I've folded out the `sameLocation` predicate now. It helped that the link @jbj sent was _exactly_ the same scenario as this one!
The performance seems to be about the same as before, though. So maybe the optimizer had a lucky join order. In any case, the query still terminates in a reasonable amount of time on Wireshark.